app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,489 @@
+    return filesForTransfer;
+  }
+
+  public void setFilesForTransfer(@NonNull ArrayList<FileItem> fileItems) {
+    this.filesForTransfer = fileItems;
+  }
+
+  public void incrementTotalFilesSent() {
+    this.filesSent++;
+  }
+
+  public boolean allFilesSent() {
+    return (filesSent == totalFilesForTransfer);
+  }
+
+  public @NonNull String getZimStorageRootPath() {
+    return (sharedPreferenceUtil.getPrefStorage() + "/Kiwix/");
+  }
+
+  public @NonNull InetAddress getFileReceiverDeviceAddress() {
+    return fileReceiverDeviceAddress;
+  }
+
+  public static void copyToOutputStream(@NonNull InputStream inputStream, @NonNull
+      OutputStream outputStream)
+      throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+  public void setClientAddress(@Nullable InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      displayToast(R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      ((Callbacks) activity).onFileTransferComplete(false);
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    if (isGroupFormed() && !isFileSender) {
+      ((Callbacks) activity).onFilesForTransferAvailable(filesForTransfer);
+
+      receiverDeviceAsyncTask = new ReceiverDeviceAsyncTask(this);
+      receiverDeviceAsyncTask.execute();
+
+    } else if (isGroupFormed()) { // && isFileSender
+      {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        fileReceiverDeviceAddress =
+            (isGroupOwner()) ? selectedPeerDeviceInetAddress : getGroupOwnerAddress();
+
+        displayToast(R.string.preparing_files, Toast.LENGTH_LONG);
+        senderDeviceAsyncTaskArray = new SenderDeviceAsyncTask(this, activity);
+        senderDeviceAsyncTaskArray.execute(fileUriArrayList.toArray(new Uri[0]));
+      }
+    }
+  }
+
+  public void changeStatus(int itemIndex, @FileItem.FileStatus int status) {
+    filesForTransfer.get(itemIndex).setFileStatus(status);
+    ((Callbacks) activity).onFileStatusChanged(itemIndex);
+  }
+
+  private void cancelAsyncTasks() {
+    if (peerGroupHandshakeAsyncTask != null) {
+      peerGroupHandshakeAsyncTask.cancel(true);
+    }
+
+    if (senderDeviceAsyncTaskArray != null) {
+      senderDeviceAsyncTaskArray.cancel(true);
+
+    } else if (receiverDeviceAsyncTask != null) {
+      receiverDeviceAsyncTask.cancel(true);
+    }
+  }
+
+  public void destroyWifiDirectManager() {
+    cancelAsyncTasks();
+
+    if (!isFileSender) {
```

Comment:
avoid using `!`, generally easier to read if something `is a thing` rather than `is not a thing`, especially when you get to an else and then you might say in your head "otherwise if this thing is not not a thing"

Reply:
So should I replace all `!<boolean>` with `<boolean> == false`?
(Same for function calls like `isLocationEnabled()` too?)

Reply:
`!` has its place but I just prefer if it can be avoided, eg
`if(boolean){
  doTheHappyPath;
}
else{
  doTheUnhappyThing;
}`

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/d78f07be6abfc481c8c46b870062346de3d0118c#diff-7e01f6f425daaca67175b74c839c6af2R378)